editor/inserter/index.js

```diff
@@ -1,18 +1,31 @@
+/**
+ * External dependencies
+ */
+import { groupBy } from 'lodash';
+
 const Inserter = () => {
 	const blocks = wp.blocks.getBlocks();
+	const blocksByCategory = groupBy( blocks, 'category' );
+	const categories = wp.blocks.getCategories();
 
 	return (
 		<div className="inserter">
 			<div className="inserter__arrow" />
 			<div className="inserter__content">
-				<div className="inserter__Category-blocks">
-					{ blocks.map( ( { slug, title, icon } ) => (
-						<div key={ slug } className="inserter__block">
-							<span className={ 'dashicons dashicons-' + icon } />
-							{ title }
+				{ categories
+					.filter( ( category ) => !! blocksByCategory[ category.id ] )
```

Comment:
I could have swear you're going to make this comment :)

More seriously, I think we have two different coding styles :). You care a lot about Performance which is good, and maybe I care a bit more about making the code meaningfull (Functional approach).

In my code we could replace the code by 

```js
categories.filter(categoryHasBlocks).map(renderCategory)
```

and with a reduce I found it less meaningfull

```js
categories.reduce(concatRenderIfHasBlocks)
```

I would use the `reduce` if we really have performance issues, but I'll choose the first approach if we don't. I hopes this explains well how I approach this kind of code